toolbox/relnotes/hierarchy.go

```diff
@@ -0,0 +1,120 @@
+package main
+
+import (
+	"fmt"
+	"os"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"github.com/google/go-github/github"
+)
+
+type dictSIG map[string]dictArea
+type dictArea map[string]dictIssue
+type dictIssue map[int]dictPR
+type dictPR map[int]bool
+
+func hierarchicalNoteLayout(f *os.File, dict dictSIG, issueMap map[int]*github.Issue) {
+	for sig, areas := range dict {
+		f.WriteString(fmt.Sprintf(" - %s\n\n", strings.Title(sig)))
+		for area, issues := range areas {
+			f.WriteString(fmt.Sprintf("    - %s\n\n", strings.Title(area)))
+			for issue, prs := range issues {
+				if issue >= 0 {
+					f.WriteString(fmt.Sprintf("        - %s (#%d)\n", *issueMap[issue].Title, issue))
+				} else {
+					f.WriteString(fmt.Sprintf("        - NullIssue\n"))
+				}
+				for pr := range prs {
+					f.WriteString(fmt.Sprintf("            * %s (#%d, @%s)\n", extractReleaseNote(issueMap[pr]), pr, *issueMap[pr].User.Login))
+				}
+				f.WriteString("\n")
+			}
+		}
+	}
+}
+
+// createHierarchicalNote given release PRs and issue map, creates hierarchical release note
+// map[SIG]map[Area]map[Issue]PR.
+func createHierarchicalNote(prs []int, issueMap map[int]*github.Issue) dictSIG {
+	var dict = dictSIG{}
+
+	for _, pr := range prs {
+		issues := extractFixedIssues(*issueMap[pr].Body)
+		if len(issues) == 0 {
+			setNoteDict(dict, "nullSig", "nullArea", -1, pr)
```

Comment:
I think we could use extractIssueSIGs(issueMap[pr]) and extractIssueArea(issueMap[pr]) instead of "nullSig" and "nullArea" to still sort these issues, since many prs don't link to an issue.

Reply:
@jennybuckley In our design doc we want the automation to enforce every release-note PR linking to at least one issue. But I agree that the rule seems to be not applied yet. I pushed an update to use SIG & area labels from PRs with no issue associated.